src/main/java/com/google/devtools/build/lib/bazel/bzlmod/VendorManager.java

```diff
@@ -83,14 +89,56 @@ public void vendorRepos(Path externalRepoRoot, ImmutableList<RepositoryName> rep
       // 3. Move the external repo to vendor dir. It's fine if this step fails or is interrupted,
       // because the marker file under external is gone anyway.
       FileSystemUtils.moveTreesBelow(repoUnderExternal, repoUnderVendor);
-      // 4. Rename to temporary marker file after the move is done.
+      // 4. Re-plant symlinks pointing a path under the external root to a relative path
+      // to make sure the vendor src keep working after being moved or output base changed
+      replantSymlinks(repoUnderVendor, externalRepoRoot);
+      // 5. Rename to temporary marker file after the move is done.
       tMarker.renameTo(markerUnderVendor);
-      // 5. Leave a symlink in external dir.
+      // 6. Leave a symlink in external dir.
       repoUnderExternal.deleteTree();
       FileSystemUtils.ensureSymbolicLink(repoUnderExternal, repoUnderVendor);
     }
   }
 
+  /**
+   * Replants the symlinks under the specified repository directory.
+   *
+   * Re-write symlinks that originally pointing to a path under the external root to a relative path pointing
+   * to an external root symlink under the vendor directory.
+   *
+   * @param repoUnderVendor The path to the repository directory under the vendor directory.
+   * @param externalRepoRoot The path to the root of external repositories.
+   * @throws IOException If an I/O error occurs while replanting the symlinks.
+   */
+  private void replantSymlinks(Path repoUnderVendor, Path externalRepoRoot) throws IOException {
+    try {
+      Collection<Path> symlinks = FileSystemUtils.traverseTree(repoUnderVendor, Path::isSymbolicLink);
+      Path externalSymlinkUnderVendor = vendorDirectory.getChild(EXTERNAL_ROOT_SYMLINK_NAME);
+      FileSystemUtils.ensureSymbolicLink(externalSymlinkUnderVendor, externalRepoRoot);
+      for (Path symlink : symlinks) {
+        PathFragment target = symlink.readSymbolicLink();
+        if (!target.startsWith(externalRepoRoot.asFragment())) {
+          // TODO: print a warning for absolute symlinks?
+          continue;
+        }
+        PathFragment newTarget = PathFragment
+            .create("../".repeat(symlink.relativeTo(vendorDirectory).segmentCount() - 1))
+            .getRelative(EXTERNAL_ROOT_SYMLINK_NAME)
+            .getRelative(target.relativeTo(externalRepoRoot.asFragment()));
+        if (OS.getCurrent() == OS.WINDOWS){
+          // On Windows, FileSystemUtils.ensureSymbolicLink always resolves paths to absolute path.
```

Comment:
That seems error-prone, could we fix its implementation instead?

Reply:
Indeed, I did a few test on Windows, it seems even long paths work with relative paths if resolved properly. But still, I prefer to implement this in a follow-up change, so that we can rollback easily if things go wrong. WDYT?

Reply:
Hmm, just find out that junctions can only be absolute paths: https://superuser.com/questions/343074/directory-junction-vs-directory-symbolic-link

Also verified on Windows
```
pcloudy@BAZEL-WINDOWS-P C:\Users\pcloudy\workdir\junction_test>mklink /J foo bar
Junction created for foo <<===>> bar

pcloudy@BAZEL-WINDOWS-P C:\Users\pcloudy\workdir\junction_test>dir
 Volume in drive C has no label.
 Volume Serial Number is 80B1-D386

 Directory of C:\Users\pcloudy\workdir\junction_test

06/17/2024  08:22 PM    <DIR>          .
06/17/2024  08:22 PM    <DIR>          ..
06/17/2024  08:22 PM    <JUNCTION>     foo [C:\Users\pcloudy\workdir\junction_test\bar]
               0 File(s)              0 bytes
               3 Dir(s)   8,777,269,248 bytes free
```

That means it's gonna be hard to fix `FileSystemUtils.ensureSymbolicLink` since if the target is a relative path pointing to a directory, then we have to decide which property to prioritize. Choosing relative path means we need admin right (or developer mode) to create actual symlink, choosing junction means we lose the relative path. And the second choice doesn't work for our use case.